resources/assets/lib/beatmap-discussions/editor.tsx

```diff
@@ -0,0 +1,467 @@
+import * as React from 'react';
+import { createEditor, Editor as SlateEditor, Element as SlateElement, Node as SlateNode, NodeEntry, Range, Text, Transforms } from 'slate';
+import { withHistory } from 'slate-history';
+import { Editable, ReactEditor, RenderElementProps, RenderLeafProps, Slate, withReact } from 'slate-react';
+import { Spinner } from 'spinner';
+import { BeatmapDiscussionReview, DocumentIssueEmbed } from '../interfaces/beatmap-discussion-review';
+import EditorDiscussionComponent from './editor-discussion-component';
+import { EditorToolbar } from './editor-toolbar';
+import { parseFromMarkdown } from './review-document';
+import { SlateContext } from './slate-context';
+
+interface CacheInterface {
+  beatmaps?: Beatmap[];
+}
+
+interface Props {
+  beatmaps: Beatmap[];
+  beatmapset: Beatmapset;
+  currentBeatmap: Beatmap;
+  currentDiscussions: BeatmapDiscussion[];
+  discussion?: BeatmapDiscussion;
+  discussions: BeatmapDiscussion[];
+  document?: string;
+  editing?: boolean;
+  editMode?: boolean;
+}
+
+interface State {
+  posting: boolean;
+  value: SlateNode[];
+}
+
+interface TimestampRange extends Range {
+  timestamp: string;
+}
+
+export default class Editor extends React.Component<Props, State> {
+  bn = 'beatmap-discussion-editor';
+  cache: CacheInterface = {};
+  emptyDocTemplate = [{children: [{text: ''}], type: 'paragraph'}];
+  slateEditor: ReactEditor;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.slateEditor = this.withNormalization(withHistory(withReact(createEditor())));
+
+    let initialValue = this.emptyDocTemplate;
+    const saved = localStorage.getItem(`newDiscussion-${this.props.beatmapset.id}`);
+    if (!props.editMode && saved) {
+      try {
+        initialValue = JSON.parse(saved);
+      } catch (error) {
+        // tslint:disable-next-line:no-console
+        console.error('invalid json in localStorage, ignoring');
+      }
+    }
+
+    this.state = {
+      posting: false,
+      value: initialValue,
+    };
+  }
+
+  blockWrapper = (children: JSX.Element) => {
+    return (
+      <div className={`${this.bn}__block`}>
+        <div className={`${this.bn}__hover-menu`} contentEditable={false}>
+          <i className='fas fa-plus-circle' contentEditable={false} />
+          <div className={`${this.bn}__menu-content`}>
+            {this.embedButton('suggestion')}
+            {this.embedButton('problem')}
+            {this.embedButton('praise')}
+          </div>
+        </div>
+        {children}
+      </div>
+    );
+  }
+
+  componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<any>, snapshot?: any): void {
+    const editing = this.props.editing ?? false;
+    if (editing !== (prevProps.editing ?? false)) {
+      this.toggleEditMode(editing);
+    }
+  }
+
+  componentWillUpdate(): void {
+    this.cache = {};
+  }
+
+  decorate = (entry: NodeEntry) => {
+    const [node, path] = entry;
+    const ranges: TimestampRange[] = [];
```

Comment:
`decorate` sounds like a generic name but the returned ranges are strictly timestamp ranges?